### PR TITLE
Added function for parameter validation

### PR DESCRIPTION
**Pull request by KevinMenden**
## PR checklist
- [X] This comment contains a description of changes (with reason)

This adds the `Validation` class in in the `lib` folder, which checks parameter for validity.

Currenty supports:
* check that types between the schema and given parameters are matching
* issues warnings for unexpected parameters
* check that a valid choice is given for `enum` fields
* check whether core nextflow flags are specifieds as in `params`

It does not currently exit the program yet, and needs some more testing, so I'M keeping this just as draft for now.

### CODE REVIEW

**Comment by KevinMenden** [comment]
Nevermind the checks in the config - just did some testing, will change back.

File: lib/Validation.groovy [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+                valid_params.push(validateParamPair(params[p.key], p, log))
+                expected_params.push(p.key)
+            }
+        }
+
+        // Check for nextflow core params and unexpected params
+        for (specified_param in specified_param_keys){
+            // nextflow params
+            if (nf_params.contains(specified_param)){
+                log.error "ERROR: That's a nextflow param!"
```
Comment (ewels) [inline lib/Validation.groovy]: Can force a system exit here I think. Happy to effectively make it an nf-core guideline that no parameter should share a name with a core Nextflow option.

File: lib/Validation.groovy [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+                }
+            }
+
+        }
+        return valid_param
+     }
+
+    private static Class determineDefaultClass(json_param){
+        def default_value = json_param.value['default']
+        def schema_param_class = default_value.getClass()
```
Comment (ewels) [inline lib/Validation.groovy]: Any reason not to use the `type` that is set in the schema? eg. https://github.com/nf-core/cageseq/blob/0e58db8cce0555684fb71ccaa8ced7c7a8fa4fdc/nextflow_schema.json#L38

File: lib/Validation.groovy (hunk 1 above)
Comment (KevinMenden) [reply lib/Validation.groovy]: yep will add it there and if we get failing params as well I guess?

File: lib/Validation.groovy (hunk 2 above)
Comment (KevinMenden) [reply lib/Validation.groovy]: not anymore ^^

File: lib/Validation.groovy (hunk 1 above)
Comment (ewels) [reply lib/Validation.groovy]: Not for failing (unexpected?) params. There can be lots of reasons for params that are not in the schema (custom cluster options for example), so they should be warning only.

Could be good to collect them in a variable so that we can print them again if the workflow exits with an error?

I did also wonder a while back about checking for the number of characters different to valid params and failing if it's too close. Or could lowercase and remove all punctuation and check too for example. Could be nice to force an exit in those cases.

File: lib/Validation.groovy [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+
+    }
+
+
+
+    /*
+    * Compare a pair of params (schema, command line) and check whether 
+    * they are valid
+    */
+     private static boolean validateParamPair(given_param, json_param, log){
```
Comment (ewels) [inline lib/Validation.groovy]:
```suggestion
    private static boolean validateParamPair(given_param, json_param, log){
```

File: nextflow.config [hunk 4]
```diff
@@ -20,7 +20,7 @@ params {
   save_reference = false
   min_aln_length = 15
   star_ignore_sjdbgtf = false
-  seq_center = false
+  seq_center = ''
```
Comment (ewels) [inline nextflow.config]: Can we revert these changes now?

File: lib/Validation.groovy [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+        def param_type = json_param.value['type']
+        def valid_param = true
+        def required = json_param.value['required']
+        def param_enum = json_param.value['enum']
+        def param_pattern = json_param.value['pattern']
+        
+        // Check only if required or parameter is given
+        if (required || given_param){
+            def given_param_class = given_param.getClass()
```
Comment (ewels) [inline lib/Validation.groovy]: Where is `getClass()` defined now?

File: lib/Validation.groovy [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+
+                switch(param_type) {
```
Comment (ewels) [inline lib/Validation.groovy]: Some funky indentation going on here I think..

File: lib/Validation.groovy [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+        // Check for nextflow core params and unexpected params
+        for (specified_param in specified_param_keys){
+            // nextflow params
+            if (nf_params.contains(specified_param)){
+                log.error "ERROR: You have overwritten the core Nextflow parameter -${specified_param} with --${specified_param}!"
+                System.exit(0)
+            }
+            // unexpected params
+            if (!expected_params.contains(specified_param)){
+                log.warn "Unexpected parameter specified: ${specified_param}"
```
Comment (ewels) [inline lib/Validation.groovy]: See https://github.com/nf-core/cageseq/pull/38#discussion_r544597458 for ideas about when we might call `System.exit()` here..

File: lib/Validation.groovy [hunk 8]
```diff
@@ -0,0 +1,115 @@
...
+        if (valid_params.contains(false)){
+            System.exit(0)
+        }
+
+        // Check for nextflow core params and unexpected params
+        for (specified_param in specified_param_keys){
+            // nextflow params
+            if (nf_params.contains(specified_param)){
+                log.error "ERROR: You have overwritten the core Nextflow parameter -${specified_param} with --${specified_param}!"
+                System.exit(0)
```
Comment (ewels) [inline lib/Validation.groovy]:
```suggestion
                System.exit(1)
```

File: lib/Validation.groovy [hunk 9]
```diff
@@ -0,0 +1,115 @@
...
+                if (!blacklist.contains(p.key)){
+                    valid_params.push(validateParamPair(params[p.key], p, log))
+                    expected_params.push(p.key)
+                }
+            }
+        }
+
+        // Exit if any invalid params where found
+        if (valid_params.contains(false)){
+            System.exit(0)
```
Comment (ewels) [inline lib/Validation.groovy]: Want to exit with a non-zero exit code here, as something went wrong..

```suggestion
            System.exit(1)
```

File: lib/Validation.groovy [hunk 10]
```diff
@@ -0,0 +1,115 @@
...
+    private static void validateParameters(params, json_schema, log, workflow){
+
+        def json = new File(json_schema).text
+        def Map json_params = (Map) new JsonSlurper().parseText(json).get('definitions')
+        def specified_param_keys = params.keySet()
+        def nf_params = ['profile', 'config', 'c', 'C', 'syslog', 'd', 'dockerize', 
+                        'bg', 'h', 'log', 'quiet', 'q', 'v', 'version']
+        def valid_params = []
+        def expected_params = []
+        def blacklist  = ['hostnames'] // ignored parameters
```
Comment (ewels) [inline lib/Validation.groovy]: Why are you skipping `hostnames`?

Also, a thread going around the internet a while back was that we should try to avoid adding in variable names like this due to `#BlackLivesMatter` awareness 🤔

File: lib/Validation.groovy [hunk 11]
```diff
@@ -0,0 +1,115 @@
...
+
+
+    /*
+    * Compare a pair of params (schema, command line) and check whether 
+    * they are valid
+    */
+     private static boolean validateParamPair(given_param, json_param, log){
+        def param_type = json_param.value['type']
+        def valid_param = true
+        def required = json_param.value['required']
```
Comment (ewels) [inline lib/Validation.groovy]: This is not set on the schema `property` level like `type` and the other keys here, but rather higher up in the schema:
https://github.com/nf-core/cageseq/blob/0e58db8cce0555684fb71ccaa8ced7c7a8fa4fdc/nextflow_schema.json#L13-L18

File: lib/Validation.groovy [hunk 12]
```diff
@@ -0,0 +1,115 @@
...
+class Validation {
+
+    /*
+    * Function to loop over all parameters defined in schema and check
+    * whether the given paremeters adhere to the specificiations
+    */
+    private static void validateParameters(params, json_schema, log, workflow){
+
+        def json = new File(json_schema).text
+        def Map json_params = (Map) new JsonSlurper().parseText(json).get('definitions')
```
Comment (ewels) [inline lib/Validation.groovy]: There may be parameters defined outside the `definitions` head key.  These are only used for grouping and you can have ungrouped params.

**Review by ewels — CHANGES_REQUESTED** [review]
Looking good, but you're starting to hit some of the JSON Schema complexities that are not obvious at first. The more I look through your code the more potential future problems I can see with various different schema structures 😓  I fought with stuff like the varying hierarchies of the `required` keyword and different structure of `definitions` and property `obejcts` for a long time over the summer.

If you can stomach it, the JSON Schema validator package should have dealt with all of this complexity for you already 😉

File: nextflow_schema.json [hunk 13]
```diff
@@ -392,6 +393,7 @@
                     "description": "Maximum amount of memory that can be requested for any single job.",
                     "default": "128.GB",
                     "fa_icon": "fas fa-memory",
+                    "pattern": "\\d+\\.?\\s*[KMGT]?B?",
```
Comment (ewels) [inline nextflow_schema.json]: As mentioned in https://github.com/nf-core/tools/issues/793, maybe we should force people to use at least GB here...

```suggestion
                    "pattern": "\\d+\\.?\\s*[GT]B",
```

File: lib/Validation.groovy (hunk 5 above)
Comment (KevinMenden) [reply lib/Validation.groovy]: that's given by groovy

File: lib/Validation.groovy (hunk 6 above)
Comment (KevinMenden) [reply lib/Validation.groovy]: heard some talk about groovy linting, guess I check that out :)

File: lib/Validation.groovy (hunk 10 above)
Comment (KevinMenden) [reply lib/Validation.groovy]: Oh shoot ...okay will change it.

Hmm blacklisted it for now because it parses as a LinkedHashMap. Will look into it.

File: lib/Validation.groovy (hunk 1 above)
Comment (KevinMenden) [reply lib/Validation.groovy]: Okay, yes I think it's good to give the user as much info as possible for debugging.
Yes a warning would be good in case it misses a valid parameter, will see about that.

**Comment by KevinMenden** [comment]
Okay, looks like there's a bit more work left to-do than I expected :-) 

Will have another go at the schema validator today, see if I can get that running and if it fixes the issues. Then try to make the code more robust.

File: nextflow.config (hunk 4 above)
Comment (KevinMenden) [reply nextflow.config]: Not sure yet, the Schema validator doesn't like it either. And it makes sense, you expect a String then it can't be a boolean.
Might be a way to workaround it - have to get warm with the library first :)

**Comment by KevinMenden** [comment]
Okay it's now working and not failing for  false-strings, memory objects and the like.
Will still have to do some cleaning and double-checking though ...

File: lib/Validation.groovy [hunk 14]
```diff
@@ -0,0 +1,105 @@
+@Grab(group='org.everit.json', module='org.everit.json.schema', version='1.5.1')
```
Comment (ewels) [inline lib/Validation.groovy]: I wonder what happens if you try to run this on a fresh system without an Internet connection 🤔

File: lib/Validation.groovy [hunk 15]
```diff
@@ -0,0 +1,105 @@
...
+        for (group in schemaParams) {
+            for (p in group.value['properties']) {
+                expectedParams.push(p.key)
+            }
+        }
+
+        for (specified_param in specified_param_keys) {
+            // nextflow params
+            if (nf_params.contains(specified_param)) {
+                log.error "ERROR: You have overwritten the core Nextflow parameter -${specified_param} with --${specified_param}!"
```
Comment (ewels) [inline lib/Validation.groovy]: Not overwritten really..

```suggestion
                log.error "ERROR: You used a core Nextflow option with two hyphens: --${specified_param}! Please resubmit with one."
```

File: lib/Validation.groovy [hunk 16]
```diff
@@ -0,0 +1,105 @@
...
+        def jsonParams = new JsonBuilder(cleanedParams)
+        JSONObject paramsJSON = new JSONObject(jsonParams.toString())
+
+        // Validate
+        try {
+            schema.validate(paramsJSON)
+        } catch (ValidationException e) {
+            log.error 'Found parameter violations!'
+            JSONObject exceptionJSON = e.toJSON()
+            printExceptions(exceptionJSON, log)
```
Comment (ewels) [inline lib/Validation.groovy]: Sys exit here?

File: lib/Validation.groovy [hunk 17]
```diff
@@ -0,0 +1,105 @@
...
+
+        for (specified_param in specified_param_keys) {
+            // nextflow params
+            if (nf_params.contains(specified_param)) {
+                log.error "ERROR: You have overwritten the core Nextflow parameter -${specified_param} with --${specified_param}!"
+                System.exit(0)
+            }
+            // unexpected params
+            if (!expectedParams.contains(specified_param)) {
+                log.warn "Unexpected parameter specified: ${specified_param}"
```
Comment (ewels) [inline lib/Validation.groovy]: Can we collect these in a map which is available to the rest of the workflow? Would be nice to repeat the log if the workflow exits with an error.

**Review by ewels — COMMENTED** [review]
Looking awesome! Much happier with this implementation, great job 😊

I wonder if we should have options to skip this function completely, and to hide warnings about unrecognised parameters (maybe specific ones - a param to specify additional expected params).

I'm also a little nervous about the library grab thing in offline environments which is something I only just thought about 😕 @pditommaso - any suggestions about importing Java libraries like this in workflows?

File: lib/Validation.groovy (hunk 14 above)
Comment (KevinMenden) [reply lib/Validation.groovy]: Good point ... that would only be relevant if someone wants to download the pipeline though, right? Otherwise, you need internet for lots of other things anyway.
But there must be some way of 'installing' it.

File: lib/Validation.groovy (hunk 16 above)
Comment (KevinMenden) [reply lib/Validation.groovy]: yep

File: lib/Validation.groovy (hunk 17 above)
Comment (KevinMenden) [reply lib/Validation.groovy]: Good idea, thanks!

**Comment by KevinMenden** [comment]
Yes, the skip function should be easy to implement. And yes it's probably nicer to omit those warnings and only show in case something goes wrong! 

I will also have to test this a bit more with more fancy pipelines before making a PR over at `tools`.
And I ended up copying the `params` anyway because that's easier to filter than a JSON object - but I'm not entirely trusting this copy yet (fear that it might be too shallow) - so I will have to do some more testing there as well.

File: lib/Validation.groovy (hunk 14 above)
Comment (pditommaso) [reply lib/Validation.groovy]: Nextflow includes Groovy [JsonSlurper](https://groovy-lang.org/json.html). Could not be used instead of downloading `org.everit.json.schema` ?

Comment (KevinMenden) [reply lib/Validation.groovy]: The idea was to use a Schema validator to avoid having to code the functionality ourselves - so the issue is not so much in parsing the JSON but in validating it against the Schema.

Comment (KevinMenden) [reply lib/Validation.groovy]: Okay so groovy stores the dependencies by default in `~/.groovy/grape` 
So if you started the workflow once online, the dependencies are saved and it works offline as well.
It is also possible to directly install dependencies to `~/.groovy/grape` using `grape install`. The directory to install the dependencies to can also be changed.

If someone runs a pipeline offline, they will usually have to run `nf-core download` before, right? So we could simply add a function to that which installs the java dependencies with grape. Should be doable. I'm testing and collecting the necessary commands to get it running ... :man_technologist:

Comment (KevinMenden) [reply lib/Validation.groovy]: Okay, all we need to do is to run these two commands when downloading the pipeline:

```
grape install org.json json [20201115]
grape install org.everit.json org.everit.json.schema [1.5.1]
```

Then it works offline, too.

**Comment by KevinMenden** [comment]
Alright did some more testing and all the different schema types seem to validate correctly.
The only thing I might add as well is to check unrecognised parameters for similarity with correct params, as you suggested Phil.

Apart from that it should be fine - let me know if you can still find some bugs :-) 

If we were to add this in the current version to tools, we would need to:

- add the groovy function to lib
- adjust the `main.nf` template to include the validation code
- add the `validate_schema` param to `nextflow.config` by default
- extend the `nf-core download` utility to get those java dependencies
- add patterns to the memory and duration objects so that they can be properly validated

I was also wondering whether we should put the complete function in `Schema.groovy`? I made a separate script mainly because it was easier to test this way .

**Comment by ewels** [comment]
> - extend the `nf-core download` utility to get those java dependencies

How would this work?

File: lib/Validation.groovy (hunk 14 above)
Comment (ewels) [reply lib/Validation.groovy]: Ah ok sorry just saw this. Great, but won't work for 99% of people using `nf-core download` sadly, as this gets it offline _on the current system_. If it's not in the resulting tarball with the pipeline files then it won't be transferred to the offline cluster.

Comment (ewels) [reply lib/Validation.groovy]: > Nextflow includes Groovy [JsonSlurper](https://groovy-lang.org/json.html). Could not be used instead of downloading `org.everit.json.schema` ?

@pditommaso we're not using this library to load JSON, we're using it to validate _JSON Schema_ - one of these things that looks simple on the face of it and turns out to be really very complex..

Comment (pditommaso) [reply lib/Validation.groovy]: I see. Then, provided this lib self-contained (ie does not pull other libraries) you can put that jar in the project `lib/` directory and NF adds it to the class path.

**Comment by KevinMenden** [comment]
It should be enough to just run this once:

```
grape install org.json json [20201115]
grape install org.everit.json org.everit.json.schema [1.5.1]
```

That installs all necessary dependencies.
Only thing I just realized now, and that might be a problem, is that grape is not installed if you just install nextflow, it comes with the complete groovy installation ... so that might be an issue, will have a closer look

**Comment by KevinMenden** [comment]
There's also the ``NXF_GRAB`` nextflow env variable, which: "Provides extra runtime dependencies downloaded from a Maven repository service."
So it might also be possible to use that for downloading the dependencies somehow.

Well I will do some testing today. If the command line version doesn't work, then a dirty workaround would be to fire up a small NF script which get's the dependencies via @Grab . 
Really not quite sure how all of this groovy magic is working :D

**Comment by KevinMenden** [comment]
Okay now I just saw your comments - continuing in the main thread so it's easier to follow.

Okay then that's a problem with downloading... also my last two comments won't help solving then I think. Not quire sure about the `NXF_GRAB` but probably that won't help either.

I'm pretty sure it pulls some other libraries unfortunately :/ 
Will try anyway. Way to go would then to download the jars in case we need them and put them to `lib/`

Considering that this parameter checking is a nice, but not a required functionality, I'm wondering whether we absolutely need this for offline use? Could just add a warning that "parameter validation not available when running offline" or something like that. Just in case we need to add too much bulk in order to get it to work.

Alright I'll try to play around with the JARs now :)

**Comment by KevinMenden** [comment]
As I feared `json-schema` needs other dependencies, and it fails because of that (otherwise imports seem to be fine).

Not sure if that can be resolved, just putting these dependencies into `lib/` as well doesn't seem to work. Will keep looking for ways to solve this.

**Comment by KevinMenden** [comment]
Okay turns out we really need those dependencies somehow because groovy fails even if we don't execute the code. It still checks for the libraries.

When specifying the dependencies with `NXF_GRAB`, nextflow dowloads them to `.nextflow/capsule/deps` and running the pipeline offline works. It's similar to specifying them with `@Grab` inside the groovy script, which saves it to `.groovy/grapes`

When `NXF_GRAB` is defined, executing any nextflow command will download the dependencies to `.nextflow/capsule/deps`. 

Currently the only way I see to make this work is to download the dependencies in one of the two ways mentioned above when running `nf-core download`, shipping it with the downloaded archive, and then telling the user to move the dependencies to the `.nextflow/capsules/deps` directory.
That's not nice because it's extra work for the user but it's the only way I can see right now, considering that putting the JAR files to `lib` didn't work.

Think I will post this to slack in the hope for more ideas :)
